core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryStream.scala

```diff
@@ -0,0 +1,146 @@
+        val markersByTimestamp = new PriorityQueueMap[MarkerKey, MsgWithTimestamp](valueOrdering = bySmallestTimestampAscending)
+        val latestMarkerSeenTimestamp = new Ref[Timestamp]
+        val latestMarkerSeenAt = new Ref[Timestamp]
+        cmd => {
+          logger.traceCommand(cmd)
+          cmd match {
+            case TickRedeliveryCommand => // nothing to do
+            case MarkerRedeliveryCommand(msg) =>
+              // update markersByTimestamp
+              msg.record.value match {
+                case _: StartMarker => markersByTimestamp.put(msg.record.key, MsgWithTimestamp(msg, msg.record.timestamp + msg.record.value.asInstanceOf[StartMarker].getRedeliverAfter)) //TODO: simplify
+                case _: EndMarker => markersByTimestamp.remove(msg.record.key)
+                case _ => throw new IllegalArgumentException()
+              }
+
+              //update latestMarkerSeen timestamps
+              latestMarkerSeenAt.update(System.currentTimeMillis())
+              if (!latestMarkerSeenTimestamp.getOption.exists(_ >= msg.record.timestamp)) {
+                latestMarkerSeenTimestamp.update(msg.record.timestamp)
+              }
+          }
+
+          // pass on all expired markers TODO: cleanup
+          val currentTime = System.currentTimeMillis()
+          val now = latestMarkerSeenAt.getOption.flatMap { lm =>
+            if (currentTime - lm < kmqConfig.getUseNowForRedeliverDespiteNoMarkerSeenForMs) {
+              /* If we've seen a marker recently, then using the latest seen marker (which is the maximum marker offset seen
+              at all) for computing redelivery. This guarantees that we won't redeliver a message for which an end marker
+              was sent, but is waiting in the topic for being observed, even though comparing the wall clock time and start
+              marker timestamp exceeds the message timeout. */
+              latestMarkerSeenTimestamp.getOption
+            } else {
+              /* If we haven't seen a marker recently, assuming that it's because all available have been observed. Hence
+              there are no delays in processing of the markers, so we can use the current time for computing which messages
+              should be redelivered. */
+              Some(currentTime)
+            }
+          }.getOrElse(currentTime)
+          logger.traceHeadOption(markersByTimestamp, now)
+
+          val toRedeliver = ArrayBuffer[CommittableMessage[MarkerKey, MarkerValue]]()
+          while (markersByTimestamp.headOption.exists(now >= _.redeliveryTime)) {
+            toRedeliver += markersByTimestamp.dequeue().msg
+          }
+          logger.traceToRedeliver(toRedeliver)
+          toRedeliver
+        }
+      }
+      .statefulMapConcat { () => // redeliver
+        val redeliverer = new RetryingRedeliverer(new DefaultRedeliverer(partition, producer, kmqConfig, kafkaClients))
```

Comment:
I suppose we can create the `redeliverer` once (outside of the stream) and  use a normal `map`? Or if resources need to be re-created on stream materialisation, then the only one that might potentially need it is the `Producer`

Reply:
Okey